src/core/experimental/array.d

```diff
@@ -0,0 +1,2299 @@
+///
+module core.experimental.array;
+
+import core.internal.traits : Unqual;
+
+// { "Imports" from Phobos
+
+// { Allocators
+
+/**
+Returns the size in bytes of the state that needs to be allocated to hold an
+object of type `T`. `stateSize!T` is zero for `struct`s that are not
+nested and have no nonstatic member variables.
+ */
+template stateSize(T)
+{
+    static if (is(T == class) || is(T == interface))
+        enum stateSize = __traits(classInstanceSize, T);
+    else static if (is(T == void))
+        enum size_t stateSize = 0;
+    else
+        enum stateSize = T.sizeof;
+}
+
+template isAbstractClass(T...)
+if (T.length == 1)
+{
+    enum bool isAbstractClass = __traits(isAbstractClass, T[0]);
+}
+
+template isInnerClass(T)
+if (is(T == class))
+{
+    static if (is(typeof(T.outer)))
+        enum isInnerClass = __traits(isSame, typeof(T.outer), __traits(parent, T));
+    else
+        enum isInnerClass = false;
+}
+
+enum classInstanceAlignment(T) = size_t.alignof >= T.alignof ? size_t.alignof : T.alignof;
+
+T emplace(T, Args...)(T chunk, auto ref Args args)
```

Comment:
Please do add a (non-ddoc) comment explaining this function, perhaps with a ref to the phobos version.